### PR TITLE
fix bcast with parallel executor + cpu

### LINKED ISSUE #12096: ParallelExecutor CPU 单机和多机精度验证
## ParallelExecutor CPU 单机和多机精度验证
验证模型： text_classification
验证代码： https://github.com/seiriosPlus/fluid_benchmark/blob/master/text_classification/train_high_api.py

### 单机: 
CPU_NUM = 4
OPTIMIZER = SGD
BatchSzie = 160
PARALLEL_EXECUTOR = False
Epoch: 0, Step: 0, Time: 0.691313028336, Loss: 0.758364140987, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.669323921204, Loss: 0.755404293537, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.492468118668, Loss: 0.75151693821, Accuracy: 0.0
Epoch: 0, Step: 3, Time: 0.634160995483, Loss: 0.74986153841, Accuracy: 0.0
Epoch: 0, Step: 4, Time: 0.559371948242, Loss: 0.746825933456, Accuracy: 0.0
Epoch: 0, Step: 5, Time: 0.65730381012, Loss: 0.744488358498, Accuracy: 0.0
Epoch: 0, Step: 6, Time: 0.57714009285, Loss: 0.740698099136, Accuracy: 0.0
Epoch: 0, Step: 7, Time: 0.565196990967, Loss: 0.737674236298, Accuracy: 0.0
Epoch: 0, Step: 8, Time: 0.670403957367, Loss: 0.735484838486, Accuracy: 0.0
Epoch: 0, Step: 9, Time: 0.60733795166, Loss: 0.732566833496, Accuracy: 0.0
Epoch: 0, Step: 10, Time: 0.618324041367, Loss: 0.72943675518, Accuracy: 0.0
Epoch: 0, Step: 11, Time: 0.654896020889, Loss: 0.726955294609, Accuracy: 0.0
Epoch: 0, Step: 12, Time: 0.606182098389, Loss: 0.724970698357, Accuracy: 0.0
Epoch: 0, Step: 13, Time: 0.535058021545, Loss: 0.720089435577, Accuracy: 0.0
Epoch: 0, Step: 14, Time: 0.530606985092, Loss: 0.717955291271, Accuracy: 0.0
Epoch: 0, Step: 15, Time: 0.55975317955, Loss: 0.716057658195, Accuracy: 0.0
Epoch: 0, Step: 16, Time: 0.57083106041, Loss: 0.713347136974, Accuracy: 0.0
Epoch: 0, Step: 17, Time: 0.583375930786, Loss: 0.710104584694, Accuracy: 0.0
Epoch: 0, Step: 18, Time: 0.625988960266, Loss: 0.70750695467, Accuracy: 0.0
Epoch: 0, Step: 19, Time: 0.561584949493, Loss: 0.705342590809, Accuracy: 0.00625000009313

- - - - - 
CPU_NUM = 4
OPTIMIZER = SGD
BatchSzie = 40 (相当于每个线程是40的BatchSize)
PARALLEL_EXECUTOR = True
Epoch: 0, Step: 0, Time: 0.363101959229, Loss: 0.758364200592, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.276148080826, Loss: 0.755404353142, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.26346206665, Loss: 0.751516878605, Accuracy: 0.0
Epoch: 0, Step: 3, Time: 0.298815965652, Loss: 0.74986153841, Accuracy: 0.0
Epoch: 0, Step: 4, Time: 0.269979953766, Loss: 0.746825933456, Accuracy: 0.0
Epoch: 0, Step: 5, Time: 0.294218063354, Loss: 0.744488418102, Accuracy: 0.0
Epoch: 0, Step: 6, Time: 0.295897960663, Loss: 0.740698099136, Accuracy: 0.0
Epoch: 0, Step: 7, Time: 0.291954994202, Loss: 0.737674236298, Accuracy: 0.0
Epoch: 0, Step: 8, Time: 0.388696908951, Loss: 0.735484838486, Accuracy: 0.0
Epoch: 0, Step: 9, Time: 0.277570962906, Loss: 0.732566773891, Accuracy: 0.0
Epoch: 0, Step: 10, Time: 0.279649972916, Loss: 0.729436695576, Accuracy: 0.0
Epoch: 0, Step: 11, Time: 0.276800870895, Loss: 0.7269551754, Accuracy: 0.0
Epoch: 0, Step: 12, Time: 0.282086849213, Loss: 0.724970698357, Accuracy: 0.0
Epoch: 0, Step: 13, Time: 0.264585971832, Loss: 0.720089375973, Accuracy: 0.0
Epoch: 0, Step: 14, Time: 0.255345106125, Loss: 0.717955350876, Accuracy: 0.0
Epoch: 0, Step: 15, Time: 0.262812137604, Loss: 0.716057658195, Accuracy: 0.0
Epoch: 0, Step: 16, Time: 0.294218063354, Loss: 0.713347196579, Accuracy: 0.0
Epoch: 0, Step: 17, Time: 0.286861896515, Loss: 0.710104644299, Accuracy: 0.0
Epoch: 0, Step: 18, Time: 0.310491085052, Loss: 0.707507014275, Accuracy: 0.0
Epoch: 0, Step: 19, Time: 0.30881690979, Loss: 0.705342590809, Accuracy: 0.00625000009313

>  多机：
CPU_NUM = 4 
Trainer*Pserver = 1*1
BatchSize = 40 (相当于每个线程是40的BatchSize)

数据详情：
Epoch: 0, Step: 0, Time: 0.645524024963, Loss: 0.758364200592, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.452592849731, Loss: 0.716291725636, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.413733005524, Loss: 0.688405871391, Accuracy: 0.75
Epoch: 0, Step: 3, Time: 0.470826148987, Loss: 0.663906574249, Accuracy: 0.75
Epoch: 0, Step: 4, Time: 0.439845085144, Loss: 0.643519043922, Accuracy: 0.75
Epoch: 0, Step: 5, Time: 0.439889907837, Loss: 0.622890949249, Accuracy: 0.75
Epoch: 0, Step: 6, Time: 0.444410085678, Loss: 0.604803562164, Accuracy: 0.75
Epoch: 0, Step: 7, Time: 0.403778076172, Loss: 0.588454723358, Accuracy: 0.75
Epoch: 0, Step: 8, Time: 0.483130931854, Loss: 0.567219376564, Accuracy: 0.75
Epoch: 0, Step: 9, Time: 0.452316999435, Loss: 0.553434789181, Accuracy: 0.75
Epoch: 0, Step: 10, Time: 0.46025800705, Loss: 0.534703552723, Accuracy: 0.75
Epoch: 0, Step: 11, Time: 0.422217130661, Loss: 0.514850497246, Accuracy: 0.75
Epoch: 0, Step: 12, Time: 0.421497106552, Loss: 0.499343156815, Accuracy: 0.75
Epoch: 0, Step: 13, Time: 0.402488946915, Loss: 0.487519323826, Accuracy: 0.75
Epoch: 0, Step: 14, Time: 0.384223937988, Loss: 0.476423442364, Accuracy: 0.75
Epoch: 0, Step: 15, Time: 0.41647195816, Loss: 0.457093030214, Accuracy: 0.75
Epoch: 0, Step: 16, Time: 0.443618059158, Loss: 0.445936471224, Accuracy: 0.75
Epoch: 0, Step: 17, Time: 0.42789888382, Loss: 0.433399736881, Accuracy: 0.75
Epoch: 0, Step: 18, Time: 0.473331212997, Loss: 0.420853078365, Accuracy: 0.75
Epoch: 0, Step: 19, Time: 0.466829061508, Loss: 0.409123629332, Accuracy: 0.75

目前1x1的精度无法对齐，原因正在查找中。

### LINKED ISSUE #12113: bcast misbehavior while dist train with parallelexe + cpu
Some misbehavior in ParallelExecutor::BCastParamsToGPUs with CPUPlace.

### LINKED ISSUE #12116: distributed AUC

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #12096 
fixed #12113 
project: #12008

local training:

CPU_NUM=8,batch_size=10
```text
Epoch: 0, Step: 0, Time: 0.281064033508, Loss: 0.699198842049, Accuracy: 0.015625
Epoch: 0, Step: 1, Time: 0.312509059906, Loss: 0.696446716785, Accuracy: 0.11562500149
Epoch: 0, Step: 2, Time: 0.309959888458, Loss: 0.693225681782, Accuracy: 0.524999976158
Epoch: 0, Step: 3, Time: 0.316581010818, Loss: 0.690290749073, Accuracy: 0.84375
Epoch: 0, Step: 4, Time: 0.327558994293, Loss: 0.687200427055, Accuracy: 0.987499952316
Epoch: 0, Step: 5, Time: 0.320170164108, Loss: 0.683837294579, Accuracy: 1.0
Epoch: 0, Step: 6, Time: 0.306921005249, Loss: 0.680849075317, Accuracy: 1.0
Epoch: 0, Step: 7, Time: 0.323282957077, Loss: 0.677796363831, Accuracy: 1.0
Epoch: 0, Step: 8, Time: 0.308493852615, Loss: 0.674684643745, Accuracy: 1.0
Epoch: 0, Step: 9, Time: 0.320489883423, Loss: 0.671876907349, Accuracy: 1.0
Epoch: 0, Step: 10, Time: 0.306677103043, Loss: 0.668930530548, Accuracy: 1.0
```

dist training

CPU_NUM=8, batch_size=10

``` text
Epoch: 0, Step: 0, Time: 0.504511833191, Loss: 0.711458265781, Accuracy: 0.0
Epoch: 0, Step: 1, Time: 0.451756000519, Loss: 0.705009400845, Accuracy: 0.0
Epoch: 0, Step: 2, Time: 0.450530052185, Loss: 0.702516257763, Accuracy: 0.0
Epoch: 0, Step: 3, Time: 0.435416936874, Loss: 0.699390649796, Accuracy: 0.015625
Epoch: 0, Step: 4, Time: 0.467040061951, Loss: 0.696782469749, Accuracy: 0.0718749985099
Epoch: 0, Step: 5, Time: 0.450443029404, Loss: 0.694233775139, Accuracy: 0.321874976158
Epoch: 0, Step: 6, Time: 0.462790966034, Loss: 0.691370487213, Accuracy: 0.734375
Epoch: 0, Step: 7, Time: 0.44303393364, Loss: 0.688292503357, Accuracy: 0.975000023842
Epoch: 0, Step: 8, Time: 0.440885066986, Loss: 0.685889422894, Accuracy: 1.0
Epoch: 0, Step: 9, Time: 0.456746101379, Loss: 0.683402359486, Accuracy: 1.0
Epoch: 0, Step: 10, Time: 0.443961143494, Loss: 0.680531382561, Accuracy: 1.0
```

### CODE REVIEW

**Comment by seiriosPlus** [comment]
loss和acc感觉还是差了一点, 需要再验证

**Comment by Yancey0623** [comment]
需要固定下Parameter的seed再验证下。

**Review by seiriosPlus — APPROVED** [review]
LGTM!

File: paddle/fluid/framework/parallel_executor.h [hunk 1]
```diff
@@ -66,7 +66,7 @@ class ParallelExecutor {
   void Run(const std::vector<std::string> &fetch_tensors,
            const std::string &fetched_var_name);
 
-  void BCastParamsToGPUs(const std::unordered_set<std::string> &vars) const;
+  void BCastParamsToDevs(const std::unordered_set<std::string> &vars) const;
```
Comment (jacquesqiao) [inline paddle/fluid/framework/parallel_executor.h:69]: 感觉叫做 BCastParamsToDevices比较容易理解

**Pull request merged**